Background tasks logs intermediate connection issues as errors

**Description**

We can see in SaaS that a lot of intermediate connections issues are logged as errors, causing to flood the error reporting.

Expectation would be here to log this as warning, as the request/connection is retried (based on the log statement at least).



[Error](https://console.cloud.google.com/logs/query;query=error_groups.id%3D%22CL7rsaOb9OHu0QE%22%0AlogName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.location%3D%22europe-north1%22%0Aresource.labels.cluster_name%3D%22worker-3%22%0Aresource.labels.project_id%3D%22camunda-saas-int%22%0Aresource.labels.pod_name%3D%22zeebe-1%22%0Aresource.labels.namespace_name%3D%22db344356-bad0-4b67-9334-92eed8f183d1-zeebe%22;pinnedLogId=2024-12-04T16:11:15.169595315Z%2Fk7doa0vxof15qhog;cursorTimestamp=2024-12-04T16:11:15.169595315Z;startTime=2024-12-04T15:42:19.537Z;endTime=2024-12-04T16:42:19.537Z?inv=1&invt=AbjSiQ&project=camunda-saas-int):

```
java.util.concurrent.CompletionException: java.net.ConnectException: Timeout connecting to [elasticsearch-es-http/10.44.45.56:9200]
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.base/java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.net.ConnectException: Timeout connecting to [elasticsearch-es-http/10.44.45.56:9200]
	at org.apache.http.nio.pool.RouteSpecificPool.timeout(RouteSpecificPool.java:169) ~[httpcore-nio-4.4.16.jar:4.4.16]
	at org.apache.http.nio.pool.AbstractNIOConnPool.requestTimeout(AbstractNIOConnPool.java:630) ~[httpcore-nio-4.4.16.jar:4.4.16]
	at org.apache.http.nio.pool.AbstractNIOConnPool$InternalSessionRequestCallback.timeout(AbstractNIOConnPool.java:896) ~[httpcore-nio-4.4.16.jar:4.4.16]
	at org.apache.http.impl.nio.reactor.SessionRequestImpl.timeout(SessionRequestImpl.java:198) ~[httpcore-nio-4.4.16.jar:4.4.16]
	at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processTimeouts(DefaultConnectingIOReactor.java:213) ~[httpcore-nio-4.4.16.jar:4.4.16]
	at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvents(DefaultConnectingIOReactor.java:158) ~[httpcore-nio-4.4.16.jar:4.4.16]
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor.execute(AbstractMultiworkerIOReactor.java:351) ~[httpcore-nio-4.4.16.jar:4.4.16]
	at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.execute(PoolingNHttpClientConnectionManager.java:221) ~[httpasyncclient-4.1.5.jar:4.1.5]
	at org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1.run(CloseableHttpAsyncClientBase.java:64) ~[httpasyncclient-4.1.5.jar:4.1.5]
	... 1 more
```

Error group: 

- https://console.cloud.google.com/errors/detail/CL7rsaOb9OHu0QE;service=zeebe;time=P7D;locations=global?inv=1&invt=AbjSiQ&project=camunda-saas-int
- https://console.cloud.google.com/errors/detail/CL7rsaOb9OHu0QE;service=zeebe;time=P7D;locations=global?inv=1&invt=AbjSiQ&project=camunda-saas-int-chaos

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/ReschedulingTask.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/ReschedulingTask.java
@@ -8,12 +8,18 @@
 package io.camunda.exporter.tasks;
 
 import io.camunda.zeebe.util.ExponentialBackoff;
+import java.net.ConnectException;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import org.slf4j.Logger;
 
 public final class ReschedulingTask implements Runnable {
+  private static final Set<Class<? extends Exception>> BACKGROUND_SUPPRESSED_EXCEPTIONS =
+      Set.of(SocketTimeoutException.class, ConnectException.class, SocketException.class);
   private final BackgroundTask task;
   private final int minimumWorkCount;
   private final ScheduledExecutorService executor;
@@ -70,13 +76,17 @@ public final class ReschedulingTask implements Runnable {
   }
 
   private long onError(final Throwable error) {
+    final String logMessage =
+        "Error occurred while performing a background task; operation will be retried";
     errorDelayMs = errorStrategy.applyAsLong(errorDelayMs);
 
-    // TODO: it's likely in some cases, we do want to log things as errors, but we need to
-    //  distinguish this from "normal" cases (e.g. missing data in ES, or ES temporarily
-    //  unavailable, etc.)
-    logger.warn(
-        "Error occurred while performing a background task; operation will be retried", error);
+    if (BACKGROUND_SUPPRESSED_EXCEPTIONS.contains(error.getCause().getClass())
+        && !logger.isDebugEnabled()) {
+      logger.warn("{}. `{}`", logMessage, error.getCause().getMessage());
+    } else {
+      logger.error(logMessage, error);
+    }
+
     return errorDelayMs;
   }
 

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/ReschedulingTask.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/ReschedulingTask.java
@@ -80,8 +80,7 @@ public final class ReschedulingTask implements Runnable {
         "Error occurred while performing a background task; operation will be retried";
     errorDelayMs = errorStrategy.applyAsLong(errorDelayMs);
 
-    if (BACKGROUND_SUPPRESSED_EXCEPTIONS.contains(error.getCause().getClass())
-        && !logger.isDebugEnabled()) {
+    if (BACKGROUND_SUPPRESSED_EXCEPTIONS.contains(error.getCause().getClass())) {
       logger.warn("{}. `{}`", logMessage, error.getCause().getMessage());
     } else {
       logger.error(logMessage, error);

